### ISSUE TITLE
Reference Key Vault secret as VM password

### ISSUE THREAD

**Issue by cognsom** [pendrica/chef-provisioning-azurerm#27]
How to mention the VM password as reference to a key Vault secret in chef recipe?

**Comment by stuartpreston** [comment]
Hi - without trying this out, I think my approach would be to take take the ID of your keyvault as a parameter, adding it as a parameter in your ARM template and specify it as such in your recipe, i.e. using it as `[parameters('keyvault_id')]` in your ARM template and 

```ruby
parameters keyvault_id: '/subscriptions/<subscription_id>/resourceGroups/<resource_group_name>/providers/Microsoft.KeyVault/vaults/<key_vault_name>'
```

**Comment by cognsom** [comment]
But my ARM Template adminPassword is a string.
"resources": [        {            "apiVersion": "2016-04-30-preview",            "name": "[parameters('virtualMachineName')]",            "type": "Microsoft.Compute/virtualMachines",            "location": "[resourceGroup().location]",            "dependsOn": [                "[concat('Microsoft.Network/networkInterfaces/', parameters('networkInterfaceName'))]"            ],            "properties": {                "osProfile": {                    "computerName": "[parameters('virtualMachineName')]",                    "adminUsername": "[parameters('adminUsername')]",                    "adminPassword": "[parameters('adminPassword')]"                },                "hardwareProfile": {                    "vmSize": "[parameters('virtualMachineSize')]"                },

How will I able to add keyvault_id and secret separately

**Comment by cognsom** [comment]
Could you please update on this asap

**Comment by stuartpreston** [comment]
The purpose of the project is to provide a way to execute ARM templates with their parameters vi Chef. So if it’s not working I’d appreciate your help to debug working out what we can improve. 

Are you able to do what you need to with an ARM template and normal parameters file? If so then we can look at what isn’t working here. 

If you can post your working ARM template and what you’ve tried. Running `Chef-client -l debug` should put the template into stdout for debugging. 

Are you trying to follow the same approach as in: https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-manager-keyvault-parameter ?

Please send some debug output so we can progress further if there is an issue.

**Comment by cognsom** [comment]
Using ARM Template and normal parameter file with keyvault is working, both the files are attached
[files.zip](https://github.com/pendrica/chef-provisioning-azurerm/files/1606431/files.zip)

In Chef Recipe, all the parameters can be passed in key value pair. So, I passed keyvault direclty as value to 'adminPassword' parameter.

require 'chef/provisioning/azurerm'
with_driver "AzureRM:<subsciption_id>"

azure_resource_template <name_> do
 resource_group <resource_group>
 template_source "#{Chef::Config[:cookbook_path]}/<json_path>"
 parameters virtualMachineName: '',
 networkInterfaceName: '',
 adminUsername: '',
 adminPassword: '{
 "id": "/subscriptions/<subscription_id>/resourceGroups/<resource_group_name>/providers/Microsoft.KeyVault/vaults/<key_vault_name>"
 },
 "secretName": "adminPassword"
 }'
 end

When i run Chef-client -l debug the deployment is **Failed.**

deployments result: #<Azure::ARM::Resources::Models::DeploymentExtended:0x0000000004d742c8 @id="/subscriptions/< >/resourceGroups/DevEnvResources/providers/Microsoft.Resources/deployments/deploymentvm", @name="deploymentvm", @properties=#<Azure::ARM::Resources::Models::DeploymentPropertiesExtended:0x0000000004d73c88 @provisioning_state="Failed", @correlation_id="db288f96-4c3a-4ad7-a5a3-e6133ea48031", @timestamp=#<DateTime: 2018-01-05T10:44:23+00:00 ((2458124j,38663s,823364000n),+0s,2299161j)>, @providers=[#<Azure::ARM::Resources::Models::Provider:0x0000000004d724a0 @namespace="Microsoft.Compute", @resource_types=[#<Azure::ARM::Resources::Models::ProviderResourceType:0x0000000004d71a50 @resource_type="virtualMachines", @locations=["northeurope"]>]>, #<Azure::ARM::Resources::Models::Provider:0x0000000004d70d58 @namespace="Microsoft.Network", @resource_types=[#<Azure::ARM::Resources::Models::ProviderResourceType:0x0000000004d70380 @resource_type="networkInterfaces", @locations=["northeurope"]>]>], @dependencies=[#<Azure::ARM::Resources::Models::Dependency:0x0000000004d53460 @depends_on=[#<Azure::ARM::Resources::Models::BasicDependency:0x0000000004d527e0 @id="/subscriptions/< >/resourceGroups/DevEnvResources/providers/Microsoft.Network/networkInterfaces/AZNEUCHEFCLIENT", @resource_type="Microsoft.Network/networkInterfaces", @resource_name="AZNEUCHEFCLIENT">, #<Azure::ARM::Resources::Models::BasicDependency:0x0000000004d519f8 @id="/subscriptions/< >/resourceGroups/analyticsenvresources/providers/Microsoft.Storage/storageAccounts/analyticsdevtorage", @resource_type="Microsoft.Storage/storageAccounts", @resource_name="analyticsdevtorage">], @id="/subscriptions/< >/resourceGroups/DevEnvResources/providers/Microsoft.Compute/virtualMachines/AZNEUCHEFCLIENT", @resource_type="Microsoft.Compute/virtualMachines", @resource_name="AZNEUCHEFCLIENT">], @parameters={"virtualMachineName"=>{"type"=>"String", "value"=>"AZNEUCHEFCLIENT"}, "adminUsername"=>{"type"=>"String", "value"=>"chefclient"}, "adminPassword"=>{"type"=>"SecureString"}, "networkInterfaceName"=>{"type"=>"String", "value"=>"AZNEUCHEFCLIENT"}}, @mode="Incremental">>

    - Resource Template deployment reached end state of 'Failed'.
    - deploy or re-deploy Resource Manager template 'deploymentvm'


Please suggest how should I modify my recipe.

**Comment by cognsom** [comment]
Could you please suggest how to mention keyvault reference in chef recipe

**Comment by stuartpreston** [comment]
I think you will have to modify your recipe to take a string parameter and pass that string parameter to the location in your template. The parameters on the resource cannot take an object.

**Comment by cognsom** [comment]
But I want to use azure keyvault instead of hard coding the secure password.

**Comment by stuartpreston** [comment]
Right, think about how you would accomplish that using a new string parameter. You’d call it keyvault_id or something, then provide the ID of the secret as the value. Then you would modify your ARM template to provide the value to the keyvault reference as parameter(‘keyvault_id’).

**Comment by cognsom** [comment]
As suggested in , https://stackoverflow.com/questions/37144593/key-vault-values-from-deployment-and-linked-templates-parameters
To check the normal way, modified the parameters file to reference the keyvault secret as shown below.
"adminPassword": {
            "value": "[reference(resourceId(parameters('keyvaultrg'), 'Microsoft.KeyVault/vaults/secrets', parameters('keyVaultName'), parameters('secretName'))).secretUri]"
        }

But the deployment fails,
Deployment failed. Correlation ID: 3189f5c0-d291-427a-a957-5635394bfb85. {
  "error": {
    "code": "InvalidParameter",
    "target": "adminPassword",
    "message": "The supplied password must be between 8-123 characters long and must satisfy at least 3 of password complexity requirements from the following: \r\n1) Contains an uppercase character\r\n2) Contains a lowercase character\r\n3) Contains a numeric digit\r\n4) Contains a special character\r\n5) Control characters are not allowed"
  }
}

**Comment by cognsom** [comment]
Please suggest were am I going wrong

**Comment by stuartpreston** [comment]
If I understand correctly, adminPassword takes either a ‘value’ or ‘reference’ element.  You need to use the latter.

**Comment by cognsom** [comment]
If adminPassword is taken as 'reference', will not be able to set adminPassword while running through Chef recipe. So, trying to get the value from keyvault directly and assign.

Please suggest any work around for this.

**Comment by stuartpreston** [comment]
I don’t think I have anything else to add. All Chef is doing is executing your ARM template. Have you tried executing this without specifying parameters, I.e hard coding the values in the ARM template. If that deploys then you can add back in the parameters. Ultimately I don’t think this is a recipe issue as all the recipe is doing is executing an ARM template with the specified string parameters.

**Comment by cognsom** [comment]
chef recipe is executing ARM Templates while passing string parameters.
But am unable to pass keyvault as parameter value in chef recipe.
In chef recipe, can we access azure keyvault to fetch the secret value and then assign it to ARM template parameters section.

**Comment by stuartpreston** [comment]
Yes you are correct in your description of this problem. Which is why I think you should be able to make this work using string parameters only to build up the required reference. 

Chef does not have a way itself of retrieving the secret value from Keyvault. For that it would need a keyvault client.

**Comment by cognsom** [comment]
Can you share any document or link related to keyvault client to retrive secret from keyvault.
